// tslint/lib/rules/noStringThrowRule.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Lint = require("tslint");
var ts = require("typescript");
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new Walker(sourceFile, this.getOptions()));
    };
    return Rule;
}(Lint.Rules.AbstractRule));
/* tslint:disable:object-literal-sort-keys */
Rule.metadata = {
    ruleName: "no-string-throw",
    description: "Flags throwing plain strings or concatenations of strings " +
        "because only Errors produce proper stack traces.",
    options: null,
    optionsDescription: "",
    type: "functionality",
    typescriptOnly: false,
};
/* tslint:enable:object-literal-sort-keys */
Rule.FAILURE_STRING = "Throwing plain strings (not instances of Error) gives no stack traces";
exports.Rule = Rule;
var Walker = (function (_super) {
    __extends(Walker, _super);
    function Walker() {
        return _super.apply(this, arguments) || this;
    }
    Walker.prototype.visitThrowStatement = function (node) {
        var expression = node.expression;
        if (this.stringConcatRecursive(expression)) {
            var fix = new Lint.Fix(Rule.metadata.ruleName, [new Lint.Replacement(expression.getStart(), expression.getEnd() - expression.getStart(), "new Error(" + expression.getText() + ")")]);
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING, fix));
        }
        _super.prototype.visitThrowStatement.call(this, node);
    };
    Walker.prototype.stringConcatRecursive = function (node) {
        switch (node.kind) {
            case ts.SyntaxKind.StringLiteral:
            case ts.SyntaxKind.NoSubstitutionTemplateLiteral:
            case ts.SyntaxKind.TemplateExpression:
                return true;
            case ts.SyntaxKind.BinaryExpression:
                var n = node;
                var op = n.operatorToken.kind;
                return op === ts.SyntaxKind.PlusToken &&
                    (this.stringConcatRecursive(n.left) ||
                        this.stringConcatRecursive(n.right));
            case ts.SyntaxKind.ParenthesizedExpression:
                return this.stringConcatRecursive(node.expression);
            default:
                return false;
        }
    };
    return Walker;
}(Lint.RuleWalker));
